### PR TITLE
Do substitution until fixed point

### LINKED ISSUE #4400: Align DEFINE and DATA_KW behavior
- [ ] Communicate to the rest of the team / ert-users on slack/yammer our intentions

Previously, DEFINES only resultet in limited support of mixing "define before use" and "use before define":

The following follows the "define before use" order (`<USER>` is defined before it is used in the definition of `<HOME>`)
```
DEFINE <USER> eja
DEFINE <HOME> /home/<USER>
```
"use before define" is simply the opposite order:
```
DEFINE <HOME> /home/<USER>
DEFINE <USER> eja
```
Mixing these two orders has limited support (you can mix up to three times for a chain of define dependencies).
Also, there is no hard requirement that a key starts `<` and ends with `>` (something we have seen in the wild). This means that substitution can lead to computationally complex behavior.

### PR DESCRIPTION
**Pull request by eivindjahren**
**Issue**
Resolves #4400 

This ensures that order of defines/data_kws do not impact substitution by doing it until fixed-point (with an upper limit on number of iterations to ensure termination).

This means that if keywords are not checked, DEFINES are very complex computationally. In order to restrict how complicated a config can be, we should give warnings if:

1. DEFINE/DATA_KW keywords do not start with `<` and end with `>`.
1. If keywords are nested, e.g. `<ENS-<ITER>>`.

If the above is not followed, there is potential for undefined behavior.

Such a check should probably be done by some mechanism for suggesting migration paths for deprecated configs, and unconventional keywords eventually resulting in hard fails.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

**Review by valentin-krasontovitsch — APPROVED** [review]
looking good!

**Comment by mortalisk** [comment]
Do we simply not allow < or > in define keywords? Or is it complete nested keywords <A> we don't allow?

**Comment by eivindjahren** [comment]
> Do we simply not allow < or > in define keywords? Or is it complete nested keywords we don't allow?

We don't allow nested, unbracketed etc. I will extend the test

File: src/clib/lib/res_util/subst_list.cpp [hunk 1]
```diff
@@ -339,8 +350,19 @@ bool subst_list_update_string(const subst_list_type *subst_list,
 char *subst_list_alloc_filtered_string(const subst_list_type *subst_list,
                                        const char *string) {
     char *filtered_string = util_alloc_string_copy(string);
-    if (subst_list)
-        subst_list_update_string(subst_list, &filtered_string);
+    if (subst_list) {
```
Comment (mortalisk) [inline src/clib/lib/res_util/subst_list.cpp:353]: Do we need to repeat this here?

Comment (mortalisk) [reply src/clib/lib/res_util/subst_list.cpp:353]: The following code that does the substitution that is.

File: tests/test_config_parsing/test_res_config.py [hunk 2]
```diff
@@ -114,3 +114,20 @@ def test_res_config_throws_on_missing_forward_model_job(config_dict):
...
+def test_that_non_bracketed_defines_warns(caplog):
+    with open("test.ert", "w", encoding="utf-8") as fh:
+        fh.write(
+            dedent(
+                """
+                NUM_REALIZATIONS  1
+                DEFINE A B
+                """
+            )
+        )
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/test_config_parsing/test_res_config.py:132]: Added!

**Review by mortalisk — APPROVED** [review]
DEFINE L ME
DEFINE G R
DEFINE T G
DEFINE M E

LGTM

**Pull request merged**